feat: switch to Azure Functions v4

derkoe opened the pull request:
This switches to the new v4 runtime of Azure functions in Azure Static Webapps

* remove function.json - it is not needed anymore
* rename function path to "sk_render" since v4 does not allow the path to start with "__"
* add @azure/functions to external esbuild options (if not already there) -  it's needed by the new runtime
    
See also: https://techcommunity.microsoft.com/t5/apps-on-azure-blog/azure-functions-version-4-of-the-node-js-programming-model-is-in/ba-p/3773541

closes #159

geoffrich commented on demo/func/host.json:
```diff
@@ -2,6 +2,6 @@
 	"extensionBundle": {
 		"id": "Microsoft.Azure.Functions.ExtensionBundle",
-		"version": "[2.*, 3.0.0)"
+		"version": "[4.0.0, 5.0.0)"
```
There's a external Azure function in this folder that also needs to be converted to the v4 model

geoffrich commented on index.js:
```diff
@@ -7,26 +7,7 @@ import esbuild from 'esbuild';
-	]
-}
-`;
+const ssrFunctionRoute = '/api/sk_render';
```
The docs need to be updated to mention that this route is now reserved https://github.com/derkoe/svelte-adapter-azure-swa/blob/7a855e8885385ba12a8a71f603dc6f5e8537b732/README.md#L122-L123

geoffrich commented on files/entry.js:
```diff
@@ -17,99 +18,101 @@ const server = new Server(manifest);
+				return ipAddress;
+			},
+			platform: {
+				user: httpRequest.user,
```
This needs a README update: https://github.com/derkoe/svelte-adapter-azure-swa/blob/7a855e8885385ba12a8a71f603dc6f5e8537b732/README.md#L227-L228

geoffrich commented on files/entry.js:
```diff
@@ -17,99 +18,101 @@ const server = new Server(manifest);
+
+		if (debug) {
+			context.log(`SK headers: ${JSON.stringify(Object.fromEntries(rendered.headers.entries()))}`);
+			context.log(`Response: ${JSON.stringify(rendered)}`);
```
Why not continue to log the response here?

geoffrich commented on files/entry.js:
```diff
@@ -17,99 +18,101 @@ const server = new Server(manifest);
- * */
-function toRequest(context) {
-	const { method, headers, rawBody, body } = context.req;
-	// because we proxy all requests to the render function, the original URL in the request is /api/__render
```
Can we keep the comment?

geoffrich commented on files/entry.js:
```diff
@@ -17,99 +18,101 @@ const server = new Server(manifest);
+		headers: new Headers(headers),
+		// @ts-ignore
+		body: httpRequest.body,
+		duplex: 'half'
```
Why set this property?

geoffrich commented on files/headers.js:
```diff
@@ -1,30 +1,35 @@
 		if (key === 'set-cookie') {
 			const cookieStrings = set_cookie_parser.splitCookiesString(value);
+			// @ts-ignore
 			resCookies.push(...set_cookie_parser.parse(cookieStrings));
```
Looks like this is due to a type mismatch on the `sameSite` property - are we sure the `sameSite` property returned by set_cookie_parser matches what Azure expects?

> Type 'string' is not assignable to type '"Strict" | "Lax" | "None"'

geoffrich reviewed (commented):
Thanks, this is looking pretty good! Just some small comments and requests for clarification. Unit tests are failing too, I think because we're writing some different files.

geoffrich commented on files/entry.js:
```diff
@@ -17,99 +18,101 @@ const server = new Server(manifest);
-		headers: new Headers(headers)
-	};
-
-	if (method !== 'GET' && method !== 'HEAD') {
-		init.body = Buffer.isBuffer(body)
-			? body
-			: typeof rawBody === 'string'
-				? Buffer.from(rawBody, 'utf-8')
-				: rawBody;
-	}
```
So this logic isn't needed any more under the v4 model?

geoffrich commented:
@derkoe no rush, but please re-request my review on the PR once you've addressed my comments - that way it doesn't get lost in my inbox

derkoe commented:
> @derkoe no rush, but please re-request my review on the PR once you've addressed my comments - that way it doesn't get lost in my inbox

Guess I've addressed all the issues - sorry for the delay.

ktarmyshov commented:
Guys, I have tried the V4 adapter, it seems to be working with the local SWA, but when deployed to the azure (github actions), deployment says - i'm done alright, but in the web I get "404 - not found". I packed the module with npm pack and reference it in my app via file:....

I have issues with logging in azure, so cannot really navigate, why it's happening.

ktarmyshov commented:
after some research, it seems that the function fails to start, but I cannot figure out how to get hands on the crash logs for the SWA. @derkoe @geoffrich if you happen to see the message and maybe suggest, what could be done, would be great. I updated all of the deps and adjusted the code accordingly, but this did not help unfortunately.

geoffrich reviewed (approved):
Thanks for the PR! Sorry it took so long to get merged.
